solidity/contracts/system/KeepFactorySelection.sol

```diff
@@ -122,7 +136,11 @@ library KeepFactorySelection {
     /// Otherwise, calls selection strategy providing addresses of both
     /// factories to make a choice. Additionally, passes the selection seed
     /// evaluated from the current request counter value.
+    /// Unless lock is set, it calls vendors to obtain the latests factories
+    /// versions.
     function refreshFactory(Storage storage _self) internal {
+        _self.keepStakeFactory = getKeepStakedFactory(_self);
```

Comment:
Actually, does it make sense to write them here? The current solution is correct bug gas-inefficient. Let's assume we have `_self.factorySelector` set. Then, for each factory selection, we perform calls to two vendors and `SSTORE` their addresses. This is not efficient, because we do not need to update `_self.keepStakeFactory` and `_self.ethStakeFactory`, we can pass the in-memory references to `self.factorySelector.selectFactory`. Why to `SSTORE` them, given that are interested in `selectedFactory` in `selectFactory`.

I suggest fetching two factories from vendors and passing then to `selectFactory` but not writing anything else in this function than `selectedFactory`. Instead, we can write them in `lockFactoriesVersions`.

Reply:
5de8998a76ab26dd5df83fbdf229a874a7430f3d